onnxruntime/core/graph/graph_utils.cc

```diff
@@ -399,48 +375,66 @@ bool IsGraphInput(const Graph& graph, const NodeArg* input) {
       const auto& graph_inputs = graph.GetInputsIncludingInitializers();
-      constant_initializer = std::none_of(graph_inputs.cbegin(), graph_inputs.cend(),
-                                          [&initializer_name](const NodeArg* input) {
-                                            return input->Name() == initializer_name;
-                                          });
+      bool is_constant = std::none_of(graph_inputs.cbegin(), graph_inputs.cend(),
+                                      [&initializer_name](const NodeArg* input) {
+                                        return input->Name() == initializer_name;
+                                      });
+
+      if (!is_constant) {
+        initializer = nullptr;
+      }
     }
   } else if (check_outer_scope && graph.IsSubgraph()) {
-    constant_initializer = IsConstantInitializer(*graph.ParentGraph(), initializer_name, check_outer_scope);
+    initializer = GetConstantInitializer(*graph.ParentGraph(), initializer_name);
   }
 
-  return constant_initializer;
+  return initializer;
+}
+
+bool IsConstantInitializer(const Graph& graph, const std::string& initializer_name, bool check_outer_scope) {
+  const onnx::TensorProto* initializer = GetConstantInitializer(graph, initializer_name, check_outer_scope);
+  return initializer != nullptr;
 }
 
 bool NodeArgIsConstant(const Graph& graph, const NodeArg& node_arg) {
   return IsConstantInitializer(graph, node_arg.Name(), true);
 }
 
-bool AllNodeInputsAreConstant(const Graph& graph, const Node& node) {
+bool AllNodeInputsAreConstant(const Graph& graph, const Node& node, InitializedTensorSet& constant_inputs) {
+  // only initializers can be constant, and there's no edge from a node to an initializer
+  // so the input edges count must be 0
   if (node.GetInputEdgesCount() > 0) {
     return false;
   }
+
   for (const auto* input_def : node.InputDefs()) {
     // Important note: when an initializer appears in the graph's input, this input will not be considered constant,
-    // because it can be overriden by the user at runtime. For constant folding to be applied, the initializer should
+    // because it can be overridden by the user at runtime. For constant folding to be applied, the initializer should
     // not appear in the graph's inputs (that is the only way to guarantee it will always be constant).
-    if (!NodeArgIsConstant(graph, *input_def)) {
+    const ONNX_NAMESPACE::TensorProto* initializer = GetConstantInitializer(graph, input_def->Name(), true);
+    if (initializer) {
+      constant_inputs.insert({input_def->Name(), initializer});
+    } else {
+      constant_inputs.clear();
       return false;
     }
   }
+
   return true;
 }
 
+void ReplaceInitializer(Graph& graph, const std::string& original_name, const ONNX_NAMESPACE::TensorProto& initializer,
+                        bool check_outer_scope) {
+  ORT_ENFORCE(ReplaceInitializerImpl(graph, original_name, initializer, check_outer_scope),
```

Comment:
Do we need the ReplaceInitializerImpl?

Reply:
I think it would be confusing to have ReplaceInitializer return bool but throw if it isn't able to replace the initializer, as that would signal to the caller they might need to check the return value.

however I need a bool return to know it whether it was replaced in an ancestor graph. 

so to make the api clear (don't need to check return value) I created a local implementation helper that returned bool.